### PR TITLE
Adding basic regrid tests from regular grid to supported grids

### PR DESCRIPTION
**Pull request by agstephens**
@ellesmith88 and @sol1105: this PR is merging my changes into the `regrid-main` branch. 

I have modified a few things in the `clisops/core/regrid.py` and `clisops/ops/regrid.py` modules.

I have also updated the tests.

The main updates are:
1. Merged in changes from `master`
2. Added working tests in `tests/ops/test_regrid.py`
3. Added dependency on a new repo that contains the grid description files for our target grids: https://github.com/roocs/roocs-grids
4. Renamed `Regridder` property of classes in `clisops/core/regrid.py` to lower-case standard: `regridder`
5. Added a `set_standard_name(coord_type)` method to the `clisops.core.regrid::Grid` class - to force any grid descriptions to have "latitude" and "longitude" standard names on the relevant coordinates (in `xarray`) so that `xESMF` knows how to handle them.

Please look through my changes and check that I have not introduced anything strange.

### CODE REVIEW

**Comment by agstephens** [comment]
@ellesmith88: the CI is failing, probably mainly due to linting issues. Could I ask you to pull the branch locally and sort out the `black` and `pep8` errors? Thanks

**Comment by ellesmith88** [comment]
@agstephens yes, will have a look now and try to sort it

File: clisops/core/regrid.py [hunk 1]
```diff
@@ -0,0 +1,1538 @@
...
+import warnings
+from pathlib import Path
+from typing import Tuple, Union
+
+import cf_xarray as cfxr
+import numpy as np
+import roocs_grids
+import scipy
+import xarray as xr
+import xesmf as xe
```
Comment (ellesmith88) [inline clisops/core/regrid.py:11]: Maybe needs something like 

```
try:
    import xesmf as xe
except ImportError:
    raise ValueError("Package xesmf >= 0.5.2 is required to use regridding")
```

File: tests/core/test_regrid.py [hunk 2]
```diff
@@ -0,0 +1,321 @@
...
+        assert w.regridder.filename == "bilinear_80x180_120x240_peri.nc"
+
+    def test_from_id(self):
+        pass
+
+    def test_from_disk(self):
+        pass
+
+
+class TestRegrid:
```
Comment (ellesmith88) [inline tests/core/test_regrid.py:305]: These tests probably need to use `load_esgf_test_data`

Comment (ellesmith88) [reply tests/core/test_regrid.py:305]: And need to not run if xesmf isn't installed

**Comment by ellesmith88** [comment]
@agstephens All the changes look fine to me. I fixed the linting and changed the scipy verison added to work with python 3.6 but looks like there is an issue with xesmf now and some test data not being available

**Comment by agstephens** [comment]
@sol1105: I think the tests are failing because we haven't installed `xesmf` properly/successfully in the pypi continuous-integration rules. Have you been able to install it successfully with `pip` or do you always install using `conda`?

**Comment by sol1105** [comment]
My procedure is installing esmpy through conda and xESMF with pip from the git repository.

I'm not sure I am grasping all that is going on in the automated checks, but
- build / pypi 3.8/3.9 are failing because xesmf is not found. I don't see it getting installed at all.
- build / conda-xesmf installs esmpy and xesmf 0.5.3 via conda and then installs a newer xesmf version with pip (0.5.4). The test fails because a data file cannot be found (.mini-esgf-data/master/test_data/badc/cmip6/data/CMIP6/CMIP/CAS/FGOALS-g3/historical/r1i1p1f1/Amon/tas/gn/v20190818/tas_Amon_FGOALS-g3_historical_r1i1p1f1_gn_185001.nc).

**Comment by ellesmith88** [comment]
@agstephens @sol1105 I don't think xesmf if fully pip installable - which is why we don't have it in the `requirements.txt` file - so we need to pass over the import and skip the tests that use it for the pypi CI. The conda-xesmf job will be able to run these tests.

File: clisops/core/regrid.py [hunk 3]
```diff
@@ -0,0 +1,1538 @@
...
+        # Test if xESMF
+        elif all([var in self.ds.coords for var in xESMF_vars]) and all(
+            [dim in self.ds.dims for dim in xESMF_dims]
+        ):
+            return "xESMF"
+
+        # Test if CF standard_names latitude and longitude can be found
+        elif (
+            get_coord_by_type(self.ds, "latitude") is not None
+            and get_coord_by_type(self.ds, "longitude") is not None
```
Comment (sol1105) [inline clisops/core/regrid.py:804]:
```suggestion
            and get_coord_by_type(self.ds, "longitude", ignore_aux_coords=False) is not None
```

File: clisops/core/regrid.py [hunk 4]
```diff
@@ -0,0 +1,1538 @@
...
+
+        # Test if xESMF
+        elif all([var in self.ds.coords for var in xESMF_vars]) and all(
+            [dim in self.ds.dims for dim in xESMF_dims]
+        ):
+            return "xESMF"
+
+        # Test if CF standard_names latitude and longitude can be found
+        elif (
+            get_coord_by_type(self.ds, "latitude") is not None
```
Comment (sol1105) [inline clisops/core/regrid.py:803]:
```suggestion
            get_coord_by_type(self.ds, "latitude", ignore_aux_coords=False) is not None
```

**Comment by sol1105** [comment]
- I am having a problem with `roocs_utils.xarray_utils.xarray_utils.get_coord_by_type`:
  ```python
  if ignore_aux_coords and coord_id not in ds.dims: 
       continue
  ```
  That condition causes that eventually None is returned as coordinate for example for the test dataset CMIP6_TOS_ONE_TIME_STEP / .mini-esgf-data/master/test_data/badc/cmip6/data/CMIP6/CMIP/MPI-M/MPI-ESM1-2-HR/historical/r1i1p1f1/Omon/tos/gn/v20190710/tos_Omon_MPI-ESM1-2-HR_historical_r1i1p1f1_gn_185001.nc .

  Also the test that uses that dataset fails because of this (test_grid_init_ds_tos_curvilinear in tests/core/test_regrid.py).
  Do you think it has any negative side effects to always call `get_coord_by_type` with `ignore_aux_coords=False`?
  The problem here are the 2D coordinate arrays for curvilinear grids to be regarded as auxiliary coordinates, so only the 1D index arrays are checked whether they are longitude/latitude arrays (and for those it is properly decided that they are not).


- For latitude/longitude bounds to not be recognized as latitude/longitude coordinates, we need cf-xarray>=0.6 where that bug has been fixed.

File: tests/_common.py [hunk 5]
```diff
@@ -117,3 +127,31 @@ def cmip6_archive_base():
...
+    "master/test_data/badc/cmip6/data/CMIP6/CMIP/MPI-M/MPI-ESM1-2-LR/historical/r1i1p1f1/Omon/tos/gn/v20190710/tos_Omon_MPI-ESM1-2-LR_historical_r1i1p1f1_gn_185001-186912.nc",
+).as_posix()
+
+
+# test daatsets used for regridding tests - one time step, full lat/lon
+# cmip6 atmosphere grid
+# x-res = 2.0
+CMIP6_TAS_ONE_TIME_STEP = Path(
+    MINI_ESGF_CACHE_DIR,
+    "master/test_data/badc/cmip6/data/CMIP6/CMIP/CAS/FGOALS-g3/historical/r1i1p1f1/Amon/tas/gn/v20190818/tas_Amon_FGOALS-g3_historical_r1i1p1f1_gn_185001.nc",
```
Comment (sol1105) [inline tests/_common.py:143]: Something goes wrong when downloading this file? The CI always crashes here.

File: environment.yml [hunk 6]
```diff
@@ -12,12 +12,14 @@ dependencies:
...
+ - xesmf>=0.6.0
  - dask>=2.6.0
  - bottleneck>=1.3.1,<1.4
  - pyproj>=2.5
  - requests>=2.0
- - roocs-utils>=0.2.1
- - cf_xarray>=0.3.1
-# - pip:
+ - roocs-utils>=0.3.0
+ - cf_xarray>=0.5.1
```
Comment (sol1105) [inline environment.yml:21]: Needs to be set to 0.6.0

File: clisops/core/regrid.py [hunk 7]
```diff
@@ -0,0 +1,1516 @@
...
+import roocs_grids
+import scipy
+import xarray as xr
+
+try:
+    import xesmf as xe
+except ImportError:
+    raise ValueError(
+        "Package xesmf >= 0.6.0 is required to use the regridding functionality."
+    )
```
Comment (sol1105) [inline clisops/core/regrid.py:17]: The xesmf import here causes problems now with the CI. How to deal with this?
In the core/__init__.py one can try to import xesmf, and if not possible, not import anything from core/regrid.py?

**Review by sol1105 — CHANGES_REQUESTED** [review]

File: clisops/core/regrid.py (hunk 7 above)
Comment (agstephens) [reply clisops/core/regrid.py]: @sol1105 I have made a change to how we manage the import of `xesmf`. It uses a decorator that we attach to the main functions and `__init__()` constructors for classes that require `xesmf`, e.g.:

```
@require_xesmf
def regrid(ds, regridder, adaptive_masking_threshold=0.5):
```

I am still trying to understand where the other failures are coming from in the CI.

File: environment.yml (hunk 6 above)
Comment (agstephens) [reply environment.yml:21]: Done

File: tests/_common.py (hunk 5 above)
Comment (agstephens) [reply tests/_common.py:143]: I have updated and this is now hopefully fixed.

**Pull request merged**